volatility/framework/plugins/windows/dumpfiles.py

```diff
@@ -0,0 +1,245 @@
+            requirements.SymbolTableRequirement(name="nt_symbols", description="Windows kernel symbols"),
+            requirements.IntRequirement(name='pid',
+                                        description="Process ID to include (all other processes are excluded)",
+                                        optional=True),
+            requirements.IntRequirement(name='virtaddr',
+                                        description="Dump a single _FILE_OBJECT at this virtual address",
+                                        optional=True),
+            requirements.IntRequirement(name='physaddr',
+                                        description="Dump a single _FILE_OBJECT at this physical address",
+                                        optional=True),
+            requirements.PluginRequirement(name='pslist', plugin=pslist.PsList, version=(1, 0, 0)),
+            requirements.PluginRequirement(name='handles', plugin=handles.Handles, version=(1, 0, 0))
+        ]
+
+    def dump_file_producer(self, file_object: interfaces.objects.ObjectInterface,
+                           memory_object: interfaces.objects.ObjectInterface,
+                           layer: interfaces.layers.DataLayerInterface,
+                           desired_file_name: str) -> str:
+        """Produce a file from the memory object's get_available_pages() interface.
+
+        :param file_object: the parent _FILE_OBJECT
+        :param memory_object: the _CONTROL_AREA or _SHARED_CACHE_MAP
+        :param layer: the memory layer to read from
+        :param desired_file_name: name of the output file
+        :return: result status
+        """
+        filedata = interfaces.plugins.FileInterface(desired_file_name)
+        try:
+            # Description of these variables:
+            #   memoffset: offset in the specified layer where the page begins
+            #   fileoffset: write to this offset in the destination file
+            #   datasize: size of the page
+            for memoffset, fileoffset, datasize in memory_object.get_available_pages():
+                data = layer.read(memoffset, datasize, pad = True)
+                filedata.data.seek(fileoffset)
+                filedata.data.write(data)
+
+            # Avoid writing files to disk if they are going to be empty or all zeros.
+            cached_length = len(filedata.data.getvalue())
+            if cached_length == 0 or filedata.data.getvalue().count(0) == cached_length:
+                result_text = "No data is cached for the file at {0:#x}".format(file_object.vol.offset)
+            else:
+                self.produce_file(filedata)
+                result_text = "Stored {}".format(filedata.preferred_filename)
+        except exceptions.InvalidAddressException:
+            result_text = "Unable to dump file at {0:#x}".format(file_object.vol.offset)
+
+        return result_text
+
+    def process_file_object(self, file_obj: interfaces.objects.ObjectInterface) -> Tuple:
```

Comment:
This definitely feels like a candidate for becoming a `classmethod`.  I think it would need `context, layer_name`.  I feel we could test that the layer is an Intel layer, and then we'd know we could get the physical layer using `contexts.layers[layer_name].config['base_layer']`.  It's slightly dangerous but far less dangerous than hard coding the layer name to pull.  For the safest option, we'd ask for two layer names to the classmethod...

Reply:
Running into some issues with this. I was attempting to go with the two layer name option where the caller passes in the primary and base layer but MHL and I were having some issues with figuring out how the caller should determine what the base layer is. config["base_layer"] is giving a key error. If you have any ideas on how the caller can determine the base layer that would be much appreciated!

Reply:
Sorry, my bad, it's not `base_layer` for intel layers, it's `memory_layer` (from https://github.com/volatilityfoundation/volatility3/blob/master/volatility/framework/layers/intel.py#L226 ).  So `memory_layer` is a configuration option of `intel` layers specifically.  Hopefully that's enough to get what you needed, but you'll need to get it from the intel layer's config, not `self.config`.